### PR TITLE
make spark-rapids-ml run on the driver without rapids libraries

### PR DESCRIPTION
**Pull request by wbo4958**
This PR wraps the rapids libraries into the TYPE_CHECKING branch for mypy check, and moves the rapids libraries imports into the code which should run on the executor.

The current auto-infer cuml parameter solution still needs to import the cuml counterpart class, which still breaks the rule that the spark-rapids-ml application should run on the driver without rapids libraries and GPU. So we need to figure out how to refactor this part.

### CODE REVIEW

**Comment by wbo4958** [comment]
build

**Review by eordentlich — COMMENTED** [review]
Good solution.

**Comment by lijinf2** [comment]
Look good.

File: python/src/spark_rapids_ml/common/cuml_context.py [hunk 1]
```diff
@@ -18,21 +18,18 @@
...
 from asyncio import AbstractEventLoop
-from typing import Any, List, Optional, Tuple
+from typing import TYPE_CHECKING, Any, List, Optional, Tuple
 
 import psutil
 
 # need this first to load shared ucx shared libraries from ucx-py instead of raft-dask
-from ucp import Endpoint
+if TYPE_CHECKING:
+    from ucp import Endpoint
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/common/cuml_context.py:27]: Good catch

File: python/src/spark_rapids_ml/common/cuml_context.py [hunk 2]
```diff
@@ -100,18 +101,22 @@ def __init__(
...
     @property
-    def handle(self) -> Optional[Handle]:
+    def handle(self) -> Optional["Handle"]:
         return self._handle
 
     def __enter__(self) -> "CumlContext":
         if not self.enable:
             return self
 
+        from raft_dask.common.nccl import nccl
```
[the remark this answers is not in the document]
Comment (wbo4958) [reply python/src/spark_rapids_ml/common/cuml_context.py:111]: Seems yes. If I disable it, I will run into below issue

``` console
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/tmp/spark-62bbdf50-6c10-422c-afa7-f4a62081b7c7/userFiles-76256b5d-75f3-4743-8b97-e248b409d6c1/sparkcuml.zip/spark_rapids_ml/core.py", line 429, in _train_udf
  File "./sparkcuml.zip/spark_rapids_ml/common/cuml_context.py", line 116, in __enter__
    self._nccl_comm = nccl()
NameError: name 'nccl' is not defined

```

**Comment by wbo4958** [comment]
build

**Comment by wbo4958** [comment]
build

**Review by eordentlich — APPROVED** [review]
👍

**Pull request merged**